File: solutions/python/grep/1/grep.py
def grep(pattern, flags, files):
    (
        prepend_line_numbers,
        only_file_names,
        is_insensitive,
        invert_match,
        match_entire_line,
    ) = (
        "-n" in flags,
        "-l" in flags,
        "-i" in flags,
        "-v" in flags,
        "-x" in flags,
    )
    pattern = pattern.lower() if is_insensitive else pattern
    more_than_one_file = len(files) > 1
    result = ""
    for file in files:
        for line_number, line in enumerate(open(file).readlines()):
            line_insensitive = line.lower() if is_insensitive else line
            if invert_match ^ (
                not match_entire_line
                and pattern in line_insensitive
                or pattern + "\n" == line_insensitive
            ):
                if only_file_names:
                    result += file + "\n"
                    break
                result += (
                    (f"{file}:" if more_than_one_file else "")
                    + (f"{line_number+1}:" if prepend_line_numbers else "")
                    + line
                )
    return result

File: solutions/go/grep/1/grep.go
package grep

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"slices"
	"strings"
)

func Search(pattern string, flags, files []string) (result []string) {
	if slices.Contains(flags, "-i") {
		pattern = strings.ToLower(pattern)
	}

	for _, path := range files {
		file, err := os.Open(path)
		if err != nil {
			panic(err)
		}
		defer file.Close()

		fileCount := len(files)
		scanner := bufio.NewScanner(file)
		lineNumber := 0
		for scanner.Scan() {
			lineNumber++
			line := scanner.Text()
			check := line
			if slices.Contains(flags, "-i") {
				check = strings.ToLower(check)
			}

			isExactSearch := slices.Contains(flags, "-x")
			isInvertedSearch := slices.Contains(flags, "-v")
			isPositiveMatch := isExactSearch && check == pattern || !isExactSearch && strings.Contains(check, pattern)
			if !isInvertedSearch && isPositiveMatch || isInvertedSearch && !isPositiveMatch {
				fileNameSuffix := ""
				if fileCount > 1 {
					fileNameSuffix = fmt.Sprintf("%s:", path)
				}
				if slices.Contains(flags, "-l") {
					result = append(result, path)
					break
				}
				if slices.Contains(flags, "-n") {
					line = fmt.Sprintf("%d:%s", lineNumber, line)
				}
				result = append(result, fmt.Sprintf("%s%s", fileNameSuffix, line))
			}
		}

		if err := scanner.Err(); err != nil {
			log.Fatal(err)
		}
	}
	return
}
